internal/testutils/daemon.go

```diff
@@ -0,0 +1,134 @@
+	"path/filepath"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+type daemonOptions struct {
+	cachePath  string
+	cacheDB    string
+	socketPath string
+}
+
+// DaemonOption represents an optional function that can be used to override some of the daemon default values.
+type DaemonOption func(*daemonOptions)
+
+// WithCachePath overrides the default cache path of the daemon.
+func WithCachePath(path string) DaemonOption {
+	return func(o *daemonOptions) {
+		o.cachePath = path
+	}
+}
+
+// WithCacheDB overrides the default cache database of the daemon.
+func WithCacheDB(db string) DaemonOption {
+	return func(o *daemonOptions) {
+		o.cacheDB = db
+	}
+}
+
+// WithSocketPath overrides the default socket path of the daemon.
+func WithSocketPath(path string) DaemonOption {
+	return func(o *daemonOptions) {
+		o.socketPath = path
+	}
+}
+
+// RunDaemon runs the daemon in a separate process and returns the socket path and a channel that will be closed when
+// the daemon stops.
+func RunDaemon(ctx context.Context, t *testing.T, execPath string, args ...DaemonOption) (socketPath string, stopped chan struct{}) {
+	t.Helper()
+
+	opts := &daemonOptions{}
+	for _, opt := range args {
+		opt(opts)
+	}
+
+	// Socket name has a maximum size, so we can't use t.TempDir() directly.
+	tempDir, err := os.MkdirTemp("", "authd-daemon4tests")
+	require.NoError(t, err, "Setup: failed to create temp dir for tests")
+	t.Cleanup(func() { os.RemoveAll(tempDir) })
+
+	if opts.cachePath == "" {
+		opts.cachePath = filepath.Join(tempDir, "cache")
+		require.NoError(t, os.MkdirAll(opts.cachePath, 0700), "Setup: failed to create cache dir")
+	}
+
+	if opts.cacheDB != "" {
+		CreateDBFromYAML(t, filepath.Join("testdata", "db", opts.cacheDB+".db.yaml"), opts.cachePath)
+	}
+
+	if opts.socketPath == "" {
+		opts.socketPath = filepath.Join(tempDir, "authd.socket")
+	}
+
+	config := fmt.Sprintf(`
+verbosity: 2
+paths:
+  cache: %s
+  socket: %s
+`, opts.cachePath, opts.socketPath)
+
+	configPath := filepath.Join(tempDir, "testconfig.yaml")
+	require.NoError(t, os.WriteFile(configPath, []byte(config), 0600), "Setup: failed to create config file for tests")
+
+	// #nosec:G204 - we control the command arguments in tests
+	cmd := exec.Command(execPath, "-c", configPath)
+	cmd.Stderr = os.Stderr
+	cmd.Env = AppendCovEnv(os.Environ())
+
+	stopped = make(chan struct{})
+	go func() {
+		require.NoError(t, cmd.Start(), "Setup: daemon should start with no error")
+		<-ctx.Done()
+		// The daemon can trigger some background tasks so, in order to stop it gracefully, we need to send either
+		// SIGTERM or SIGINT to tell it that it's time to cleanup and stop.
+		require.NoError(t, cmd.Process.Signal(os.Signal(syscall.SIGTERM)), "Teardown: Failed to send signal to stop daemon")
+		require.NoError(t, cmd.Wait(), "Teardown: daemon should stop with no error")
+		close(stopped)
+	}()
+
+	// Give some time for the daemon to start.
+	time.Sleep(time.Second)
```

Comment:
Wouldn't be better here to return a string channel (containing the socket path) that is filled only once the daemon has really started?

As sleeping is just not very reliable.

Reply:
So, i've looked a bit on this and I think it's easy enough to avoid unsafe timeouts, since we can just do something like:

```diff
diff --git a/internal/testutils/daemon.go b/internal/testutils/daemon.go
index 1bc1828..0591ef0 100644
--- a/internal/testutils/daemon.go
+++ b/internal/testutils/daemon.go
@@ -11,6 +11,9 @@ import (
 	"time"
 
 	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/credentials/insecure"
 )
 
 type daemonOptions struct {
@@ -108,8 +111,18 @@ paths:
 		t.Logf("Daemon stopped (%v)\n ##### STDOUT #####\n %s \n ##### END #####", err, out)
 	}()
 
-	// Give some time for the daemon to start.
-	time.Sleep(time.Second)
+	// Wait for the daemon to start.
+	conn, err := grpc.Dial("unix://"+opts.socketPath,
+		grpc.WithTransportCredentials(insecure.NewCredentials()))
+	require.NoError(t, err)
+	defer conn.Close()
+
+	waitCtx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	for conn.GetState() != connectivity.Ready {
+		conn.WaitForStateChange(waitCtx, conn.GetState())
+		require.NoError(t, waitCtx.Err())
+	}
 
 	return opts.socketPath, stopped
 }
```

And we are sure that we don't depend on random wait times that may lead to failures on some slow builders (as can be launchpad at times or autopkgtests).

Indeed this could be moved into a goroutine that is resolved by a socket path channel, but I don't think it's really needed, at the end it's fine if the test code is waiting, since we'd still have to wait at given point.